packages/vite/src/node/plugins/asset.ts

```diff
@@ -353,11 +353,22 @@ async function fileToBuiltUrl(
     const { search, hash } = parseUrl(id)
     const postfix = (search || '') + (hash || '')
     const output = config.build?.rollupOptions?.output
-    const assetFileNames =
+
+    const defaultAssetFileNames = path.posix.join(
+      config.build.assetsDir,
+      '[name].[hash][extname]'
+    )
+    // Steps to determine which assetFileNames will be actually used.
+    // First, if output is an object or string, use assetFileNames in it.
+    // And a default assetFileNames as fallback.
+    let assetFileNames: Exclude<OutputOptions['assetFileNames'], undefined> =
       (output && !Array.isArray(output) ? output.assetFileNames : undefined) ??
-      // defaults to '<assetsDir>/[name].[hash][extname]'
-      // slightly different from rollup's one ('assets/[name]-[hash][extname]')
-      path.posix.join(config.build.assetsDir, '[name].[hash][extname]')
+      defaultAssetFileNames
+    if (output && Array.isArray(output)) {
+      // Second, if output is an array, adopt assetFileNames in the first object.
+      assetFileNames = output[0].assetFileNames ?? assetFileNames
+    }
```

Comment:
I think this logic should avoid affect lib mode. use the follow logic will generate multi bundle
```js
rollupOptions: {
  external: ['vue'],
  output: [
    {
      globals: { vue: 'Vue' },
      entryFileNames: `assets/subdir/[name].js`,
      chunkFileNames: `assets/subdir/[name].js`
    },
    {
      globals: { vue: 'Vue' },
      entryFileNames: `assets/anotherSubdir/[name].js`,
      chunkFileNames: `assets/anotherSubdir/[name].js`
    }
  ]
}
```